MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,1775 @@
+        }
+        it = p.second;
+
+      }
+
+      sv0_match_sharedjettotrack_which_trkidx.push_back(sv0_match_sharedjettotrack_which_temp_trkidx);
+
+      // start collecting shared tracks of sv1 for each shared jet
+      std::multimap<int, size_t> sv1_m;
+      for (size_t k = 0; k < sv1_match_tracktojet_which_jetidx.size(); k++)
+        if (sv1_match_tracktojet_which_jetidx[k] == jet_index) { sv1_m.insert({ sv1_match_tracktojet_which_jetidx[k], k }); }
+
+      for (auto it = sv1_m.begin(); it != sv1_m.end(); )
+      {
+        auto p = sv1_m.equal_range(it->first);
+
+        while (p.first != p.second)
+        {
+
+          sv1_match_sharedjettotrack_which_temp_trkidx.push_back(sv1_match_tracktojet_which_trkidx[p.first++->second]);
+        }
+        it = p.second;
+
+      }
+
+      sv1_match_sharedjettotrack_which_trkidx.push_back(sv1_match_sharedjettotrack_which_temp_trkidx);
+
+      if (sv0_match_sharedjettotrack_which_temp_trkidx.size() == 1) {
+        sv0_lonesharedtrack_which_trkidx.push_back(sv0_match_sharedjettotrack_which_temp_trkidx[0]);
+      }
+      if (sv1_match_sharedjettotrack_which_temp_trkidx.size() == 1) {
+        sv1_lonesharedtrack_which_trkidx.push_back(sv1_match_sharedjettotrack_which_temp_trkidx[0]);
+      }
+
+      sv0_match_sharedjettotrack_which_temp_trkidx = {};
+      sv1_match_sharedjettotrack_which_temp_trkidx = {};
+
+    }
+    sv_lonesharedtrack_which_trkidx.push_back(sv0_lonesharedtrack_which_trkidx);
+    sv_lonesharedtrack_which_trkidx.push_back(sv1_lonesharedtrack_which_trkidx);
+  }
+  bool shared_jet_specialcase = (sv0_lonesharedtrack_which_trkidx.size() != 0 || sv1_lonesharedtrack_which_trkidx.size() != 0);
+  return std::pair<bool, std::vector<std::vector<size_t>>>(
+      shared_jet_specialcase,
+      sv_lonesharedtrack_which_trkidx);
+}
+
+bool MFVVertexer::match_track_jet(const reco::Track& tk, const pat::Jet& matchjet, const pat::JetCollection& jets, const int& idx) {
+
+  if (verbose) {
+    std::cout << "jet track matching..." << std::endl;
+    std::cout << "  target track pt " << tk.pt() << " eta " << tk.eta() << " phi " << tk.phi() << std::endl;
+  }
+
+  double match_thres = 1.3;
+  int jet_index = 255;
+  for (size_t j = 0; j < jets.size(); ++j) {
+    for (size_t idau = 0, idaue = jets[j].numberOfDaughters(); idau < idaue; ++idau) {
+      const reco::Candidate* dau = jets[j].daughter(idau);
+      if (dau->charge() == 0)
+        continue;
+      const reco::Track * jtk = 0;
+      const reco::PFCandidate * pf = dynamic_cast<const reco::PFCandidate*>(dau);
+      if (pf) {
+        const reco::TrackRef& r = pf->trackRef();
+        if (r.isNonnull())
+          jtk = &*r;
+      }
+      else {
+        const pat::PackedCandidate* pk = dynamic_cast<const pat::PackedCandidate*>(dau);
+        if (pk && pk->charge() && pk->hasTrackDetails())
+          jtk = &pk->pseudoTrack();
+      }

```

Comment:
Maybe this was even my idea, but why do we need to use these two different collections? Was it that the PFCandidates were only available / saved above a certain pT threshold or something of that nature? It would be good to add a comment about this.

Reply:
I think the answer is here:
https://github.com/DisplacedVertices/cmssw-usercode/blob/clean_vertexing/MFVNeutralino/plugins/EventProducer.cc#L403-L407

We might want to copy the comment over, or at least add a comment to look at the relevant code in EventProducer? I can take care of this.

Reply:
Done in 012eb470